### DISCUSSION TITLE
How to get the file in a repository with its created and last modified time

### DISCUSSION THREAD

**Discussion by tisonkun** [GitoxideLabs/gitoxide#1395] [Q&A] [answered]
I found the issue at https://github.com/korandoru/hawkeye/pull/150.

But in case of you'd like to discuss here, I repost the question:

I'd like to get the file in a repository with its created and last modified time. I suppose if we can list commits at a path, this should be implemented.

This is possible for `git` with `git log --pretty="format:%ai" /pathspec`.

But it's a bit challenging to select the proper gix features and methods to implement this.

**Comment by tisonkun** [comment]
Here is how a library use JGit for achieving this:

```java
  private RevWalk getGitRevWalk(String repoRelativePath, boolean oldestCommitsFirst) throws IOException {
    DiffConfig diffConfig = repository.getConfig().get(DiffConfig.KEY);

    RevWalk walk = new RevWalk(repository);
    walk.markStart(walk.parseCommit(repository.resolve(Constants.HEAD)));
    walk.setTreeFilter(AndTreeFilter.create(Arrays.asList(
        PathFilter.create(repoRelativePath),
        FollowFilter.create(repoRelativePath, diffConfig), // Allows us to follow files as they move or are renamed
        TreeFilter.ANY_DIFF)
    ));
    walk.setRevFilter(MaxCountRevFilter.create(checkCommitsCount));
    walk.setRetainBody(false);
    if (oldestCommitsFirst) {
      walk.sort(RevSort.REVERSE);
    }

    return walk;
  }
```

**Comment by Byron** [comment]
There is nothing to facilitate this at the moment, but I hope one day that will change.

Meanwhile, there is [example code](https://github.com/Byron/gitoxide/blob/2a9c178326b7f13ba6bc1f89fc2b9d9facbecf48/gitoxide-core/src/query/engine/update.rs#L209) that diffs the tree of each commit with the previous one, if it's not a merge-commit, and records all obtained changes in a sqlite database. From there it's easy to query, such as with:

```
> ein t query trace-path Cargo.toml
 17:40:58 db cache up to date
 17:40:58 run sql query done 1.0 round in 0.03s (36.0 round/s)
           | 2023-11-19 | bb3224c25 Δ Cargo.toml
+         | 2023-11-10 | 8f795e8ab Δ Cargo.toml
+         | 2023-10-20 | b65a80b05 Δ Cargo.toml
+         | 2023-10-15 | f34f46a38 Δ Cargo.toml
[..]
```

The same code could be used to do the same computation on the fly. It stays expensive though, there is nothing one can do about it except for attempting to not do any unnecessary computation.

**Reply by tisonkun** [reply, marked as answer]
Thank you! I implement it in https://github.com/korandoru/hawkeye/pull/150/files/e9bac68f659b60d4879d77b9f998e70df1ae8165..7de3c3566f4b3176e712318bd57af11349b63e83

Welcome to give a review and see if we can improve the code further.
